invoices/invoiceregistry.go

```diff
@@ -1043,3 +972,147 @@ func (i *InvoiceRegistry) HodlUnsubscribeAll(subscriber chan<- interface{}) {
 
 	delete(i.hodlReverseSubscriptions, subscriber)
 }
+
+// updateResult is the result of the invoice update call.
+type updateResult uint8
+
+const (
+	resultReplayToCanceled updateResult = iota
+	resultReplayToAccepted
+	resultReplayToSettled
+	resultInvoiceAlreadyCanceled
+	resultAmountTooLow
+	resultExpiryTooSoon
+	resultDuplicateToAccepted
+	resultDuplicateToSettled
+	resultAccepted
+	resultSettled
+)
+
+// String returns a human-readable representation of the invoice update result.
+func (u updateResult) String() string {
+	switch u {
+
+	case resultReplayToCanceled:
+		return "replayed htlc to canceled invoice"
+
+	case resultReplayToAccepted:
+		return "replayed htlc to accepted invoice"
+
+	case resultReplayToSettled:
+		return "replayed htlc to settled invoice"
+
+	case resultInvoiceAlreadyCanceled:
+		return "invoice already canceled"
+
+	case resultAmountTooLow:
+		return "amount too low"
+
+	case resultExpiryTooSoon:
+		return "expiry too soon"
+
+	case resultDuplicateToAccepted:
+		return "accepting duplicate payment to accepted invoice"
+
+	case resultDuplicateToSettled:
+		return "accepting duplicate payment to settled invoice"
+
+	case resultAccepted:
+		return "accepted"
+
+	case resultSettled:
+		return "settled"
+
+	default:
+		return "unknown"
+	}
+}
+
+// invoiceUpdateCtx is an object that describes the context for the invoice
+// update to be carried out.
+type invoiceUpdateCtx struct {
+	circuitKey           channeldb.CircuitKey
+	amtPaid              lnwire.MilliSatoshi
+	expiry               uint32
+	currentHeight        int32
+	finalCltvRejectDelta int32
+}
+
+// update is a callback for DB.UpdateInvoice that contains the invoice
+// settlement logic.
+func updateInvoice(ctx *invoiceUpdateCtx, inv *channeldb.Invoice) (
+	*channeldb.InvoiceUpdateDesc, updateResult) {
+
+	// Don't update the invoice when this is a replayed htlc.
+	htlc, ok := inv.Htlcs[ctx.circuitKey]
+	if ok {
+		switch htlc.State {
+		case channeldb.HtlcStateCanceled:
+			return nil, resultReplayToCanceled
+
+		case channeldb.HtlcStateAccepted:
+			return nil, resultReplayToAccepted
+
+		case channeldb.HtlcStateSettled:
+			return nil, resultReplayToSettled
+		}
+
+		panic("unexpected htlc state")
```

Comment:
Why is this being upgraded from an error to a panic?
More for my own understanding of when we panic vs error in lnd.

Reply:
It is a good question. I am not too sure myself either. If you for example do an index out of bounds, there will be a panic. This is to me kind of a similar situation. But if we want to be more defensive, we can return an error.

Reply:
What I mean is, why is it changed from an error to a panic in this PR? 
Because in master it just returns an [error](https://github.com/lightningnetwork/lnd/blob/cb2b8b451326d79808697af0d7f1d8d56e97dc07/invoices/invoiceregistry.go#L462)? 

Unless this branch is based on something other than master? Or I'm comparing this to the wrong code?

Reply:
There was no reason for the `error` return value anymore except for this. Then the question is, is it worth the extra code?

Reply:
Brought back the `error` and leaving this discussion for another time.